### ISSUE TITLE
Prevent Adam optimizer from changing dict grads

### ISSUE THREAD

**Issue by danbraunai-apollo** [ApolloResearch/e2e_sae#44]
As noticed by [Bricken et al](https://transformer-circuits.pub/2023/monosemantic-features#appendix-autoencoder-optimization), when restricting the decoder weights to have unit norm, the grads from Adam will change the norms of the dictionary elements.

I think using this format instead will fix this issue:
```
def forward(self, x: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
    c = self.encoder(x)

    # Normalize decoder weights without modifying them directly
    normalized_weights = nn.functional.normalize(self.decoder.weight, dim=0)
    
    # Use normalized weights for the forward computation
    x_hat = F.linear(c, normalized_weights, self.decoder.bias)
    return x_hat, c
```

We can add a property to our SAE class to get the normalized dictionary elements:
```
@property
def dict_elements(self):
    return nn.functional.normalize(self.decoder.weight, dim=0)
```

Nix brought up that it's possible the self.decoder.weight will grow very large over training with this method. If so we still may want to do the self.decoder.weight.data = normalize(...) after a gradient step. But I'd first want to check exactly how doing this effects the computational graph, and also maybe do a run without this and see if the decoder weight vectors grow large anyway.

Regardless, this would save us the pretty dirty computation of this gradient parallelism trick.
